File: results/2026-04-08_192624/tasks/17-artifact-cleanup-script/default-sonnet/generated-code/artifact_cleanup.py
"""
Artifact Cleanup Script
=======================
Applies retention policies to a list of build artifacts and produces a
deletion plan.  Three policy types are supported:

  max_age_days               – delete artifacts older than N days
  max_total_size_bytes       – delete oldest artifacts until total ≤ limit
  keep_latest_n_per_workflow – for each workflow, keep only the N most recent
                               workflow runs' artifacts

The script can be run in dry-run mode; in that mode it produces the same plan
output but takes no destructive action (and is clearly labeled as a dry run).

Exit codes:
  0 – success
  1 – fatal error (bad fixture file, invalid JSON, etc.)

TDD notes
---------
  CYCLE 1 GREEN: implemented parse_date + apply_age_policy
  CYCLE 2 GREEN: implemented apply_size_policy
  CYCLE 3 GREEN: implemented apply_keep_latest_n_policy
  CYCLE 4 GREEN: implemented apply_retention_policies (chains the three above)
  CYCLE 5 GREEN: implemented generate_deletion_plan + format_plan
  REFACTOR: extracted parse_date as a module-level helper; unified date
            handling; added ARTIFACT_CLEANUP_SUMMARY marker line.
"""

import argparse
import json
import sys
from datetime import datetime, timedelta, timezone
from typing import Dict, List, Optional, Set


# ─────────────────────────────────────────────────────────────
# Internal helpers
# ─────────────────────────────────────────────────────────────

def parse_date(date_str: str) -> datetime:
    """Parse an ISO-8601 date string into a timezone-aware datetime.

    Accepts the trailing 'Z' shorthand for UTC as well as explicit offsets.
    """
    # Normalise the 'Z' suffix that GitHub and many CI tools emit.
    normalised = date_str.replace("Z", "+00:00")
    return datetime.fromisoformat(normalised)


# ─────────────────────────────────────────────────────────────
# Policy functions
# Each function returns a *set of artifact names* that violate the policy.
# ─────────────────────────────────────────────────────────────

def apply_age_policy(
    artifacts: List[Dict],
    max_age_days: int,
    reference_date: datetime,
) -> Set[str]:
    """Return the set of artifact names older than *max_age_days*.

    The cutoff is calculated as reference_date − max_age_days.  Artifacts
    created strictly before the cutoff are marked for deletion; those created
    on or after are retained.
    """
    cutoff = reference_date - timedelta(days=max_age_days)
    return {
        a["name"]
        for a in artifacts
        if parse_date(a["created_at"]) < cutoff
    }


def apply_size_policy(
    artifacts: List[Dict],
    max_total_size_bytes: int,
) -> Set[str]:
    """Return the minimum set of artifact names to delete to stay under the
    total-size limit.

    Artifacts are removed oldest-first (by creation date) until the remaining
    total is at or below *max_total_size_bytes*.
    """
    total = sum(a["size"] for a in artifacts)
    if total <= max_total_size_bytes:
        return set()

    # Sort oldest first so we evict the least-valuable artifacts first.
    sorted_arts = sorted(artifacts, key=lambda a: parse_date(a["created_at"]))
    to_delete: Set[str] = set()
    for art in sorted_arts:
        if total <= max_total_size_bytes:
            break
        to_delete.add(art["name"])
        total -= art["size"]
    return to_delete


def apply_keep_latest_n_policy(
    artifacts: List[Dict],
    keep_latest_n: int,
) -> Set[str]:
    """Return artifact names that belong to workflow runs outside the top-N.

    Grouping is done by (workflow_name, workflow_run_id).  Within each
    workflow, runs are ranked by the *latest* artifact creation date in that
    run.  The N most-recent runs are kept; all others are deleted.
    """
    # Build: { workflow_name → { run_id → [artifacts] } }
    by_workflow: Dict[str, Dict[str, List[Dict]]] = {}
    for art in artifacts:
        wf = art.get("workflow_name", "default")
        run_id = art["workflow_run_id"]
        by_workflow.setdefault(wf, {}).setdefault(run_id, []).append(art)

    to_delete: Set[str] = set()
    for wf_name, runs in by_workflow.items():
        def run_date(run_id: str) -> datetime:
            # A run's "date" is the latest creation date among its artifacts.
            return max(parse_date(a["created_at"]) for a in runs[run_id])

        sorted_run_ids = sorted(runs.keys(), key=run_date, reverse=True)
        # Anything beyond the first keep_latest_n runs is evicted.
        for run_id in sorted_run_ids[keep_latest_n:]:
            for art in runs[run_id]:
                to_delete.add(art["name"])

    return to_delete


# ─────────────────────────────────────────────────────────────
# Orchestration
# ─────────────────────────────────────────────────────────────

def apply_retention_policies(
    artifacts: List[Dict],
    policies: Dict,
    reference_date: Optional[datetime] = None,
) -> Dict:
    """Apply all configured retention policies and return a result dict.

    Policies are applied in this order so that later policies operate on the
    *already-culled* set, avoiding double-counting:
      1. max_age_days
      2. max_total_size_bytes
      3. keep_latest_n_per_workflow

    Returns:
        {
            "to_delete": [artifact, ...],
            "to_retain": [artifact, ...],
            "reasons":   { artifact_name: [reason_str, ...] },
        }
    """
    if reference_date is None:
        reference_date = datetime.now(timezone.utc)

    delete_names: Set[str] = set()
    reasons: Dict[str, List[str]] = {}

    # ── Phase 1: age ──────────────────────────────────────────
    if "max_age_days" in policies:
        age_limit = policies["max_age_days"]
        for name in apply_age_policy(artifacts, age_limit, reference_date):
            delete_names.add(name)
            reasons.setdefault(name, []).append(
                f"older than {age_limit} days"
            )

    # ── Phase 2: total size ───────────────────────────────────
    if "max_total_size_bytes" in policies:
        remaining = [a for a in artifacts if a["name"] not in delete_names]
        size_limit = policies["max_total_size_bytes"]
        for name in apply_size_policy(remaining, size_limit):
            delete_names.add(name)
            reasons.setdefault(name, []).append(
                f"total size exceeds {size_limit} bytes"
            )

    # ── Phase 3: keep-latest-N ────────────────────────────────
    if "keep_latest_n_per_workflow" in policies:
        remaining = [a for a in artifacts if a["name"] not in delete_names]
        n = policies["keep_latest_n_per_workflow"]
        for name in apply_keep_latest_n_policy(remaining, n):
            delete_names.add(name)
            reasons.setdefault(name, []).append(
                f"exceeds keep-latest-{n} runs per workflow"
            )

    return {
        "to_delete": [a for a in artifacts if a["name"] in delete_names],
        "to_retain": [a for a in artifacts if a["name"] not in delete_names],
        "reasons": reasons,
    }


# ─────────────────────────────────────────────────────────────
# Plan generation & formatting
# ─────────────────────────────────────────────────────────────

def generate_deletion_plan(result: Dict, dry_run: bool = False) -> Dict:
    """Build a structured deletion plan from a retention-policy result."""
    to_delete = result["to_delete"]
    to_retain  = result["to_retain"]
    reasons    = result["reasons"]

    space_bytes = sum(a["size"] for a in to_delete)
    space_mb    = round(space_bytes / (1024 * 1024), 2)

    return {
        "dry_run": dry_run,
        "to_delete": [
            {
                "name":            a["name"],
                "size":            a["size"],
                "created_at":      a["created_at"],
                "workflow_run_id": a["workflow_run_id"],
                "reasons":         reasons.get(a["name"], []),
            }
            for a in to_delete
        ],
        "to_retain": [
            {
                "name":            a["name"],
                "size":            a["size"],
                "created_at":      a["created_at"],
                "workflow_run_id": a["workflow_run_id"],
            }
            for a in to_retain
        ],
        "summary": {
            "total_artifacts":     len(to_delete) + len(to_retain),
            "artifacts_to_delete": len(to_delete),
            "artifacts_to_retain": len(to_retain),
            "space_reclaimed_bytes": space_bytes,
            "space_reclaimed_mb":    space_mb,
        },
    }


def format_plan(plan: Dict) -> str:
    """Return a human-readable string representation of a deletion plan.

    The last section always contains a machine-readable ARTIFACT_CLEANUP_SUMMARY
    line that the test harness uses to assert exact expected values without
    having to parse prose output.

    Line format:
        ARTIFACT_CLEANUP_SUMMARY: total=N deleted=N retained=N \
            space_bytes=N space_mb=F dry_run=true|false
    """
    s = plan["summary"]
    dr = plan["dry_run"]
    lines: List[str] = []

    if dr:
        lines.append("=== DRY RUN — no artifacts will be deleted ===")
    else:
        lines.append("=== ARTIFACT CLEANUP PLAN ===")

    lines.append(f"Total artifacts  : {s['total_artifacts']}")
    lines.append(f"To delete        : {s['artifacts_to_delete']}")
    lines.append(f"To retain        : {s['artifacts_to_retain']}")
    lines.append(
        f"Space to reclaim : {s['space_reclaimed_mb']} MB"
        f" ({s['space_reclaimed_bytes']} bytes)"
    )
    lines.append("")

    if plan["to_delete"]:
        lines.append("ARTIFACTS TO DELETE:")
        for a in plan["to_delete"]:
            reasons_str = "; ".join(a["reasons"]) or "no reason specified"
            lines.append(
                f"  - {a['name']}"
                f" ({a['size']} bytes, created {a['created_at']})"
                f" [reason: {reasons_str}]"
            )

    if plan["to_retain"]:
        lines.append("")
        lines.append("ARTIFACTS TO RETAIN:")
        for a in plan["to_retain"]:
            lines.append(f"  - {a['name']} ({a['size']} bytes)")

    # Machine-readable summary — always the last line.
    lines.append("")
    lines.append(
        f"ARTIFACT_CLEANUP_SUMMARY:"
        f" total={s['total_artifacts']}"
        f" deleted={s['artifacts_to_delete']}"
        f" retained={s['artifacts_to_retain']}"
        f" space_bytes={s['space_reclaimed_bytes']}"
        f" space_mb={s['space_reclaimed_mb']}"
        f" dry_run={'true' if dr else 'false'}"
    )

    return "\n".join(lines)


# ─────────────────────────────────────────────────────────────
# CLI entry point
# ─────────────────────────────────────────────────────────────

def main() -> int:
    parser = argparse.ArgumentParser(
        description="Apply retention policies to a list of build artifacts."
    )
    parser.add_argument(
        "--fixture",
        required=True,
        help="Path to a JSON file containing artifact list and policies.",
    )
    parser.add_argument(
        "--dry-run",
        action="store_true",
        help="Show the deletion plan without taking any action.",
    )
    parser.add_argument(
        "--output-json",
        action="store_true",
        help="Print the deletion plan as JSON (default: human-readable).",
    )
    parser.add_argument(
        "--reference-date",
        help="Override today's date for age calculations (ISO-8601).",
    )
    args = parser.parse_args()

    # ── Load fixture ──────────────────────────────────────────
    try:
        with open(args.fixture) as fh:
            data = json.load(fh)
    except FileNotFoundError:
        print(f"ERROR: fixture file not found: {args.fixture}", file=sys.stderr)
        return 1
    except json.JSONDecodeError as exc:
        print(f"ERROR: invalid JSON in fixture file: {exc}", file=sys.stderr)
        return 1

    artifacts = data.get("artifacts", [])
    policies  = data.get("policies",  {})

    if not isinstance(artifacts, list):
        print("ERROR: 'artifacts' must be a JSON array.", file=sys.stderr)
        return 1
    if not isinstance(policies, dict):
        print("ERROR: 'policies' must be a JSON object.", file=sys.stderr)
        return 1

    # ── Resolve reference date ────────────────────────────────
    reference_date: Optional[datetime] = None
    if args.reference_date:
        reference_date = parse_date(args.reference_date)
    elif "reference_date" in data:
        reference_date = parse_date(data["reference_date"])

    # ── Apply policies & generate plan ────────────────────────
    result = apply_retention_policies(artifacts, policies, reference_date)
    plan   = generate_deletion_plan(result, dry_run=args.dry_run)

    if args.output_json:
        print(json.dumps(plan, indent=2))
    else:
        print(format_plan(plan))

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: results/2026-04-17_004319/tasks/17-artifact-cleanup-script/typescript-bun-sonnet46-1m-medium/generated-code/artifact-cleanup.ts
// Artifact Cleanup Script
// Applies retention policies to a list of artifacts and generates a deletion plan.
// Supports: max age, max total size, keep-latest-N per workflow, dry-run mode.

import * as fs from "fs";
import * as path from "path";

// --- Types ---

export interface Artifact {
  name: string;
  size: number; // bytes
  createdAt: Date;
  workflowRunId: string;
}

// Raw artifact as read from JSON (dates are strings)
interface ArtifactJSON {
  name: string;
  size: number;
  createdAt: string;
  workflowRunId: string;
}

export interface RetentionPolicy {
  maxAgeDays?: number;
  maxTotalSizeBytes?: number;
  keepLatestNPerWorkflow?: number;
}

export interface DeletionPlan {
  toDelete: Artifact[];
  toKeep: Artifact[];
  totalSpaceReclaimed: number;
  summary: {
    totalArtifacts: number;
    artifactsToDelete: number;
    artifactsToKeep: number;
    spaceReclaimedBytes: number;
    spaceReclaimedMB: number;
    dryRun: boolean;
  };
}

// --- Core logic ---

export function applyRetentionPolicies(
  artifacts: Artifact[],
  policy: RetentionPolicy,
  dryRun: boolean = false,
  now: Date = new Date()
): DeletionPlan {
  // Use a Set of artifact names to track which are marked for deletion.
  // Names are assumed unique in the input list.
  const toDeleteNames = new Set<string>();

  // 1. Age policy: delete artifacts older than maxAgeDays
  if (policy.maxAgeDays !== undefined) {
    const cutoffMs = now.getTime() - policy.maxAgeDays * 24 * 60 * 60 * 1000;
    for (const artifact of artifacts) {
      if (artifact.createdAt.getTime() < cutoffMs) {
        toDeleteNames.add(artifact.name);
      }
    }
  }

  // 2. Keep-latest-N per workflow: for each workflowRunId, keep only the N newest
  if (policy.keepLatestNPerWorkflow !== undefined) {
    const byWorkflow = new Map<string, Artifact[]>();
    for (const artifact of artifacts) {
      const group = byWorkflow.get(artifact.workflowRunId) ?? [];
      group.push(artifact);
      byWorkflow.set(artifact.workflowRunId, group);
    }

    for (const [, group] of byWorkflow) {
      // Sort descending by creation date — newest first
      const sorted = [...group].sort(
        (a, b) => b.createdAt.getTime() - a.createdAt.getTime()
      );
      // Everything beyond index N-1 is excess
      for (let i = policy.keepLatestNPerWorkflow; i < sorted.length; i++) {
        toDeleteNames.add(sorted[i].name);
      }
    }
  }

  // 3. Max total size: after the above deletions, if remaining size still exceeds
  //    the budget, delete oldest remaining artifacts until under budget.
  if (policy.maxTotalSizeBytes !== undefined) {
    const remaining = artifacts.filter((a) => !toDeleteNames.has(a.name));
    const totalSize = remaining.reduce((sum, a) => sum + a.size, 0);

    if (totalSize > policy.maxTotalSizeBytes) {
      // Oldest first — we delete from the oldest end to preserve recent artifacts
      const sorted = [...remaining].sort(
        (a, b) => a.createdAt.getTime() - b.createdAt.getTime()
      );
      let currentSize = totalSize;
      for (const artifact of sorted) {
        if (currentSize <= policy.maxTotalSizeBytes) break;
        toDeleteNames.add(artifact.name);
        currentSize -= artifact.size;
      }
    }
  }

  const toDelete = artifacts.filter((a) => toDeleteNames.has(a.name));
  const toKeep = artifacts.filter((a) => !toDeleteNames.has(a.name));
  const totalSpaceReclaimed = toDelete.reduce((sum, a) => sum + a.size, 0);
  const spaceReclaimedMB =
    Math.round((totalSpaceReclaimed / (1024 * 1024)) * 100) / 100;

  return {
    toDelete,
    toKeep,
    totalSpaceReclaimed,
    summary: {
      totalArtifacts: artifacts.length,
      artifactsToDelete: toDelete.length,
      artifactsToKeep: toKeep.length,
      spaceReclaimedBytes: totalSpaceReclaimed,
      spaceReclaimedMB,
      dryRun,
    },
  };
}

export function formatDeletionPlan(plan: DeletionPlan): string {
  const { summary } = plan;
  const lines: string[] = [];

  lines.push("=== Artifact Cleanup Plan ===");
  if (summary.dryRun) {
    lines.push("[DRY RUN] No artifacts will actually be deleted");
  }
  lines.push("");
  lines.push(`Total artifacts: ${summary.totalArtifacts}`);
  lines.push(`Artifacts to delete: ${summary.artifactsToDelete}`);
  lines.push(`Artifacts to keep: ${summary.artifactsToKeep}`);
  lines.push(
    `Space to reclaim: ${summary.spaceReclaimedMB.toFixed(2)} MB (${summary.spaceReclaimedBytes} bytes)`
  );
  lines.push("");

  if (plan.toDelete.length > 0) {
    lines.push("Artifacts marked for deletion:");
    for (const a of plan.toDelete) {
      const mb = (a.size / (1024 * 1024)).toFixed(2);
      lines.push(
        `  - ${a.name} | ${mb} MB | workflow: ${a.workflowRunId} | created: ${a.createdAt.toISOString()}`
      );
    }
  } else {
    lines.push("No artifacts to delete.");
  }

  if (plan.toKeep.length > 0) {
    lines.push("");
    lines.push("Artifacts retained:");
    for (const a of plan.toKeep) {
      const mb = (a.size / (1024 * 1024)).toFixed(2);
      lines.push(`  - ${a.name} | ${mb} MB | workflow: ${a.workflowRunId}`);
    }
  }

  return lines.join("\n");
}

// --- JSON deserialization helper ---

export function parseArtifacts(raw: ArtifactJSON[]): Artifact[] {
  return raw.map((r) => ({
    name: r.name,
    size: r.size,
    createdAt: new Date(r.createdAt),
    workflowRunId: r.workflowRunId,
  }));
}

// --- CLI entry point ---

if (import.meta.main) {
  const args = process.argv.slice(2);

  function getFlag(flag: string): string | undefined {
    const idx = args.indexOf(flag);
    return idx !== -1 ? args[idx + 1] : undefined;
  }

  const fixturesFile = getFlag("--fixtures");
  const policyFile = getFlag("--policy");
  const dryRun = args.includes("--dry-run");

  if (!fixturesFile || !policyFile) {
    console.error(
      "Usage: bun run artifact-cleanup.ts --fixtures <file> --policy <file> [--dry-run]"
    );
    process.exit(1);
  }

  // Resolve paths relative to CWD
  const fixturesPath = path.resolve(process.cwd(), fixturesFile);
  const policyPath = path.resolve(process.cwd(), policyFile);

  if (!fs.existsSync(fixturesPath)) {
    console.error(`Error: fixtures file not found: ${fixturesPath}`);
    process.exit(1);
  }
  if (!fs.existsSync(policyPath)) {
    console.error(`Error: policy file not found: ${policyPath}`);
    process.exit(1);
  }

  let rawArtifacts: ArtifactJSON[];
  let policy: RetentionPolicy;

  try {
    rawArtifacts = JSON.parse(fs.readFileSync(fixturesPath, "utf-8"));
  } catch (e) {
    console.error(`Error: failed to parse fixtures file: ${(e as Error).message}`);
    process.exit(1);
  }

  try {
    policy = JSON.parse(fs.readFileSync(policyPath, "utf-8"));
  } catch (e) {
    console.error(`Error: failed to parse policy file: ${(e as Error).message}`);
    process.exit(1);
  }

  const artifacts = parseArtifacts(rawArtifacts);
  const plan = applyRetentionPolicies(artifacts, policy, dryRun);
  const output = formatDeletionPlan(plan);

  console.log(output);
  console.log("");
  console.log("JSON Summary:");
  console.log(JSON.stringify(plan.summary, null, 2));
}
